atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCqlClusterContainer.java

```diff
@@ -0,0 +1,147 @@
+                    log.warn("Failed to close CQL Cluster after reloading server list", e);
+                }
+            }
+        });
+    }
+
+    public static ReloadingCqlClusterContainer of(
+            CassandraClusterConfig cassandraClusterConfig,
+            Refreshable<CassandraServersConfig> refreshableCassandraServersConfig,
+            Namespace namespace) {
+        return of(cassandraClusterConfig, refreshableCassandraServersConfig, namespace, CqlCluster::create);
+    }
+
+    public static ReloadingCqlClusterContainer of(
+            CassandraClusterConfig cassandraClusterConfig,
+            Refreshable<CassandraServersConfig> refreshableCassandraServersConfig,
+            Namespace namespace,
+            CqlClusterFactory cqlClusterFactory) {
+        return new ReloadingCqlClusterContainer(
+                cassandraClusterConfig, refreshableCassandraServersConfig, namespace, cqlClusterFactory);
+    }
+
+    /**
+     * Synchronized: See {@link #close()}.
+     */
+    private synchronized CqlCluster createNewCluster(
+            CassandraClusterConfig cassandraClusterConfig,
+            CassandraServersConfig cassandraServersConfig,
+            Namespace namespace,
+            CqlClusterFactory cqlClusterFactory) {
+        if (isClosed) {
+            throw new SafeIllegalStateException(
+                    "Attempted to create a new cluster after the container was closed. If this happens repeatedly,"
+                            + " this is likely a bug in closing the container. Otherwise, it is highly likely that the"
+                            + " container was closed at the same time as the server list was updated. If so, this error"
+                            + " can be ignored.",
+                    UnsafeArg.of("keyspace", namespace));
+        }
+        return cqlClusterFactory.create(cassandraClusterConfig, cassandraServersConfig, namespace);
+    }
+
+    /**
+     * Synchronized: A lock is taken out to ensure no new CQL Clusters are created after retrieving the current stored
+     * cql cluster to close. By doing so, we avoid closing a cluster and subsequently creating a new one that is
+     * never closed.
+     */
+    @Override
+    public synchronized void close() throws IOException {
+        isClosed = true;
+        refreshableSubscriptionDisposable.dispose();
+        Optional<CqlCluster> maybeCqlClusterToClose = currentCqlCluster.get();
+        if (maybeCqlClusterToClose.isPresent()) {
+            maybeCqlClusterToClose.get().close();
+        }
+    }
+
+    /**
+     * Gets the latest CqlCluster that reflects any changes in the server list, provided {@link #close()} has not
+     * been called.
+     *
+     * The CQL Cluster returned will be closed after {@link #close} is called, or the server list is refreshed, even
+     * if the CQL Cluster is in active use.
+     */
+    @Override
+    public CqlCluster get() {
+        return refreshableCqlCluster.get();
+    }
```

Comment:
do consumers currently handle a closed cluster? Should we be returning optionals that consumers `orElseThrow` to make this explicit?

Reply:
A very reasonable suggestion. On the other hand, the container can close a cluster that's already been retrieved (and wrapping it in Optional wouldn't change that, since the optional already has the cluster), and so consumers would _still_ need to be able to deal with a closed cluster.